pkg/syncer/cnsoperator/controller/csinodetopology/csinodetopology_controller.go

```diff
@@ -0,0 +1,336 @@
+		if err != nil {
+			msg := fmt.Sprintf("failed to fetch topology information for the nodeVM with ID %q", nodeID)
+			log.Errorf("%s. Error: %v", msg, err)
+			_ = updateCRStatus(ctx, r, instance, csinodetopologyv1alpha1.CSINodeTopologyError, msg)
+			return reconcile.Result{RequeueAfter: timeout}, nil
+		}
+
+		// Raise error if nodeVM does not have a topology label associated with each category in the
+		// vSphere config secret `Labels` section. For now, as we support only zone, region, hardcoded the value to 2
+		// TODO: Count the number of topology categories given and verify against that number.
+		if len(topologyLabels) != 2 {
+			msg := fmt.Sprintf("Detected a topology aware cluster. However, nodeVM with ID %q does not have a "+
+				"topology label for each category mentioned under the vSphere CSI config secret `Labels` section.", nodeID)
+			log.Error(msg)
+			_ = updateCRStatus(ctx, r, instance, csinodetopologyv1alpha1.CSINodeTopologyError, msg)
+			return reconcile.Result{RequeueAfter: timeout}, nil
+		}
+
+		// Update CSINodeTopology instance
+		instance.Status.TopologyLabels = topologyLabels
+		err = updateCRStatus(ctx, r, instance, csinodetopologyv1alpha1.CSINodeTopologySuccess, "")
+		if err != nil {
+			return reconcile.Result{RequeueAfter: timeout}, nil
+		}
+	} else {
+		msg := "missing zone or region information in vSphere CSI config `Labels` section"
+		log.Error(msg)
+		_ = updateCRStatus(ctx, r, instance, csinodetopologyv1alpha1.CSINodeTopologyError, msg)
+		return reconcile.Result{}, errors.New(msg)
+	}
+
+	// On successful event, remove instance from backOffDuration
+	backOffDurationMapMutex.Lock()
+	delete(backOffDuration, instance.Name)
+	backOffDurationMapMutex.Unlock()
+	log.Infof("Successfully updated topology labels for nodeVM with ID %q", nodeID)
+	return reconcile.Result{}, nil
+}
+
+func updateCRStatus(ctx context.Context, r *ReconcileCSINodeTopology, instance *csinodetopologyv1alpha1.CSINodeTopology,
+	status csinodetopologyv1alpha1.CRDStatus, failureEventMessage string) error {
+	log := logger.GetLogger(ctx)
+
+	instance.Status.Status = status
+	switch status {
+	case csinodetopologyv1alpha1.CSINodeTopologySuccess:
+		// Reset error message if previously populated
+		instance.Status.ErrorMessage = failureEventMessage
+		// Record an event on the CR
+		r.recorder.Event(instance, corev1.EventTypeNormal, "CSINodeTopologySucceeded",
+			fmt.Sprintf("Topology labels updated for nodeVM %q", instance.Spec.NodeID))
+	case csinodetopologyv1alpha1.CSINodeTopologyError:
+		instance.Status.ErrorMessage = failureEventMessage
+		// Record an event on the CR
+		r.recorder.Event(instance, corev1.EventTypeWarning, "CSINodeTopologyFailed", failureEventMessage)
+	}
+
+	// Update CSINodeTopology instance
+	err := r.client.Update(ctx, instance)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to update the CSINodeTopology instance with name %q.", instance.Spec.NodeID)
```

Comment:
Can replace this with `log.LogNewErrorf`

Reply:
LogNewErrorf creates an error right? I just need to use the message to log and update the CR

Reply:
I think thats `LogNewErrorCodef`

Reply:
Even `LogNewErrorf` returns an error.
Refer https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/3bfe9fdb1e3dec04a574f587fda18b166932bdca/pkg/csi/service/logger/logger.go#L109